crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,355 @@
+        let fetch_queue_rx = Arc::new(tokio::sync::Mutex::new(fetch_queue_rx));
+
+        let state = Arc::new(Mutex::new(State {
+            ops: HashSet::new(),
+            cool_down_list: CoolDownList::new(config.cool_down_interval_ms),
+        }));
+
+        let mut fetch_tasks = Vec::new();
+        for _ in 0..config.parallel_request_count {
+            let task = tokio::task::spawn(Inner::fetch_task(
+                state.clone(),
+                fetch_queue_tx.clone(),
+                fetch_queue_rx.clone(),
+                peer_store.clone(),
+                space_id.clone(),
+                transport.clone(),
+            ));
+            fetch_tasks.push(task);
+        }
+
+        Self {
+            state,
+            fetch_queue_tx,
+            fetch_tasks,
+        }
+    }
+
+    async fn fetch_task(
+        state: Arc<Mutex<State>>,
+        fetch_request_tx: Sender<FetchRequest>,
+        fetch_request_rx: Arc<tokio::sync::Mutex<Receiver<FetchRequest>>>,
+        peer_store: peer_store::DynPeerStore,
+        space_id: SpaceId,
+        transport: DynTransport,
+    ) {
+        while let Some((op_id, agent_id)) =
+            fetch_request_rx.lock().await.recv().await
+        {
+            let is_agent_cooling_down = {
+                let mut state_lock = state.lock().unwrap();
+
+                // Do nothing if op id is no longer in the set of ops to fetch.
+                if !state_lock.ops.contains(&(op_id.clone(), agent_id.clone()))
+                {
+                    continue;
+                }
+
+                state_lock.cool_down_list.is_agent_cooling_down(&agent_id)
+            };
+
+            // Send request if agent is not on cool-down list.
+            if !is_agent_cooling_down {
+                // Send fetch request to agent.
+                let peer = match peer_store.get(agent_id.clone()).await {
+                    Ok(Some(peer)) => match &peer.url {
+                        Some(url) => url.clone(),
+                        None => {
+                            tracing::warn!(
+                                "agent {agent_id} no longer in peer store"
+                            );
+                            continue;
+                        }
+                    },
+                    Ok(None) => {
+                        tracing::warn!(
+                            "could not find agent id {agent_id} in peer store"
+                        );
+                        continue;
+                    }
+                    Err(err) => {
+                        tracing::warn!("could not get agent id {agent_id} from peer store: {err}");
+                        continue;
+                    }
+                };
+
+                let mut data = bytes::BytesMut::new();
+                data.put(op_id.clone().0 .0);
+                data.put(agent_id.clone().0 .0);
+                let data = data.freeze();
```

Comment:
```suggestion
                // Using this simple "encoding" for now.
                // We can evaluate in the future if we need
                // anything else more complicated.
                let mut data = bytes::BytesMut::new();
                data.put(op_id.clone().0 .0);
                data.put(agent_id.clone().0 .0);
                let data = data.freeze();
```

Reply:
Actually, now that I've commented on this, I'm not sure this is sufficient. The kitsune api doesn't provide any requirements that the ids are the same length, so we can't decode this with just this information.

Reply:
Yes, ideally we use serializable types where we're sending messages like this I think.

I was just looking for this code to say, can we make sure we make the "ops to get" part of the request a list. Then it won't be a breaking protocol change if we want to batch the gets

Reply:
Using protobuf, that is?

Reply:
I think that'd be the best option yes, just to start a protocol definition for the fetch module

Reply:
To clarify, I didn't think of this quick'n dirty way as the final api to construct the data to pass in =D

Protobuf definition done here https://github.com/holochain/kitsune2/pull/35/commits/e27beedb5f7fb7fec536df84eb300b9115f6e58e